File: link_checker.cs
// link_checker.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Threading;
using System.Diagnostics;

class RedirectStep
{
    public string Url { get; set; }
    public int StatusCode { get; set; }
    public Dictionary<string, string> Headers { get; set; }
}

class Result
{
    public string Url { get; set; }
    public string FinalUrl { get; set; }
    public int StatusCode { get; set; }
    public List<RedirectStep> Redirects { get; set; } = new List<RedirectStep>();
    public double Time { get; set; }
    public long Size { get; set; }
    public string Error { get; set; }
}

class LinkChecker
{
    private HttpClient client;
    private int maxRedirects;
    private int timeout;

    public LinkChecker(int timeoutSec, int maxRedirects, Dictionary<string, string> headers)
    {
        this.timeout = timeoutSec;
        this.maxRedirects = maxRedirects;
        var handler = new HttpClientHandler()
        {
            AllowAutoRedirect = false,
        };
        client = new HttpClient(handler);
        client.Timeout = TimeSpan.FromSeconds(timeoutSec);
        foreach (var h in headers)
            client.DefaultRequestHeaders.Add(h.Key, h.Value);
    }

    public async Task<Result> CheckUrl(string url)
    {
        var stopwatch = Stopwatch.StartNew();
        var result = new Result { Url = url };

        string currentUrl = url;
        int redirectCount = 0;

        try
        {
            while (redirectCount <= maxRedirects)
            {
                var response = await client.GetAsync(currentUrl);
                int status = (int)response.StatusCode;
                result.Redirects.Add(new RedirectStep
                {
                    Url = currentUrl,
                    StatusCode = status,
                    Headers = response.Headers.ToDictionary(h => h.Key, h => string.Join(",", h.Value))
                });

                if (status >= 300 && status < 400)
                {
                    var location = response.Headers.Location?.ToString();
                    if (string.IsNullOrEmpty(location))
                        break;
                    // Разрешаем относительный URL
                    var baseUri = new Uri(currentUrl);
                    var resolved = new Uri(baseUri, location);
                    currentUrl = resolved.ToString();
                    redirectCount++;
                    if (redirectCount > maxRedirects)
                    {
                        result.Error = $"Слишком много редиректов (> {maxRedirects})";
                        break;
                    }
                }
                else
                {
                    result.StatusCode = status;
                    result.FinalUrl = currentUrl;
                    var content = await response.Content.ReadAsByteArrayAsync();
                    result.Size = content.Length;
                    break;
                }
                response.Dispose();
            }
            if (redirectCount > maxRedirects && result.Error == null)
                result.Error = $"Превышен лимит редиректов ({maxRedirects})";
        }
        catch (Exception ex)
        {
            result.Error = ex.Message;
        }

        result.Time = stopwatch.Elapsed.TotalSeconds;
        return result;
    }

    public async Task<Dictionary<string, Result>> CheckMultiple(List<string> urls, int workers)
    {
        var results = new Dictionary<string, Result>();
        var sem = new SemaphoreSlim(workers);
        var tasks = urls.Select(async url =>
        {
            await sem.WaitAsync();
            try
            {
                var r = await CheckUrl(url);
                lock (results) results[url] = r;
            }
            finally { sem.Release(); }
        });
        await Task.WhenAll(tasks);
        return results;
    }
}

class Program
{
    static void PrintResults(Dictionary<string, Result> results)
    {
        foreach (var kv in results)
        {
            var url = kv.Key;
            var r = kv.Value;
            Console.WriteLine($"\n\u001B[36m🔍 Проверка: {url}\u001B[0m");
            if (!string.IsNullOrEmpty(r.Error))
            {
                Console.WriteLine($"\u001B[31m❌ Ошибка: {r.Error}\u001B[0m");
                continue;
            }
            string statusColor = "\u001B[32m";
            string statusText = $"{r.StatusCode} OK";
            if (r.StatusCode >= 300 && r.StatusCode < 400)
            {
                statusColor = "\u001B[33m";
                statusText = $"{r.StatusCode} Redirect";
            }
            else if (r.StatusCode >= 400)
            {
                statusColor = "\u001B[31m";
                statusText = $"{r.StatusCode} Error";
            }
            Console.WriteLine($"{statusColor}✅ {r.FinalUrl}\u001B[0m");
            Console.WriteLine($"   Статус: {statusColor}{statusText}\u001B[0m");
            Console.WriteLine($"   Время: {r.Time:F2} сек");
            Console.WriteLine($"   Размер: {r.Size} байт");
            if (r.Redirects.Count > 0)
            {
                Console.WriteLine($"   Редиректы: {r.Redirects.Count}");
                for (int i = 0; i < r.Redirects.Count; i++)
                    Console.WriteLine($"     {i+1}. {r.Redirects[i].Url} -> {r.Redirects[i].StatusCode}");
            }
        }
    }

    static void SaveJSON(Dictionary<string, Result> results, string filename)
    {
        var json = JsonSerializer.Serialize(results, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
        Console.WriteLine($"\u001B[32m💾 Сохранено JSON: {filename}\u001B[0m");
    }

    static void SaveCSV(Dictionary<string, Result> results, string filename)
    {
        var sb = new StringBuilder();
        sb.AppendLine("URL,Final URL,Status,Time (s),Size (bytes),Redirects,Error");
        foreach (var kv in results)
        {
            var r = kv.Value;
            sb.AppendLine($"{kv.Key},{r.FinalUrl ?? ""},{r.StatusCode},{r.Time:F3},{r.Size},{r.Redirects.Count},{r.Error ?? ""}");
        }
        File.WriteAllText(filename, sb.ToString());
        Console.WriteLine($"\u001B[32m💾 Сохранено CSV: {filename}\u001B[0m");
    }

    static async Task Main(string[] args)
    {
        var urls = new List<string>();
        string filePath = null;
        string outputJson = null;
        string outputCsv = null;
        int workers = 10;
        int timeout = 10;
        int maxRedirects = 10;
        var headers = new Dictionary<string, string>();

        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == "--file" || args[i] == "-f") filePath = args[++i];
            else if (args[i] == "--output-json") outputJson = args[++i];
            else if (args[i] == "--output-csv") outputCsv = args[++i];
            else if (args[i] == "--workers" || args[i] == "-w") workers = int.Parse(args[++i]);
            else if (args[i] == "--timeout") timeout = int.Parse(args[++i]);
            else if (args[i] == "--max-redirects") maxRedirects = int.Parse(args[++i]);
            else if (args[i] == "--header" || args[i] == "-H")
            {
                var h = args[++i];
                var parts = h.Split(':', 2);
                if (parts.Length == 2) headers[parts[0].Trim()] = parts[1].Trim();
            }
            else if (!args[i].StartsWith("-"))
                urls.Add(args[i]);
        }

        if (filePath != null)
        {
            var lines = File.ReadAllLines(filePath);
            foreach (var line in lines)
                if (!string.IsNullOrWhiteSpace(line)) urls.Add(line.Trim());
        }

        if (urls.Count == 0)
        {
            Console.WriteLine("❌ Нет URL для проверки.");
            return;
        }

        Console.WriteLine("\u001B[36m🔗 Link Checker (C#)\u001B[0m");

        var checker = new LinkChecker(timeout, maxRedirects, headers);
        var results = await checker.CheckMultiple(urls, workers);
        PrintResults(results);

        if (outputJson != null) SaveJSON(results, outputJson);
        if (outputCsv != null) SaveCSV(results, outputCsv);
    }
}

File: link_checker.php
<?php
// link_checker.php — PHP версия

require_once 'vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Psr7\Uri;
use GuzzleHttp\Psr7\UriResolver;

class LinkChecker
{
    private $client;
    private $maxRedirects;
    private $timeout;

    public function __construct($timeout = 10, $maxRedirects = 10, $headers = [])
    {
        $this->timeout = $timeout;
        $this->maxRedirects = $maxRedirects;
        $this->client = new Client([
            'timeout' => $timeout,
            'allow_redirects' => false,
            'headers' => $headers,
            'http_errors' => false,
        ]);
    }

    public function checkUrl($url)
    {
        $start = microtime(true);
        $result = [
            'url' => $url,
            'final_url' => null,
            'status_code' => null,
            'redirects' => [],
            'time' => 0,
            'size' => 0,
            'error' => null,
        ];

        $currentUrl = $url;
        $redirectCount = 0;

        try {
            while ($redirectCount <= $this->maxRedirects) {
                $response = $this->client->get($currentUrl);
                $status = $response->getStatusCode();
                $result['redirects'][] = [
                    'url' => $currentUrl,
                    'status_code' => $status,
                    'headers' => $response->getHeaders(),
                ];

                if ($status >= 300 && $status < 400) {
                    $location = $response->getHeaderLine('Location');
                    if (!$location) break;
                    // Разрешаем относительный URL
                    $base = new Uri($currentUrl);
                    $resolved = UriResolver::resolve($base, new Uri($location));
                    $currentUrl = (string) $resolved;
                    $redirectCount++;
                    if ($redirectCount > $this->maxRedirects) {
                        $result['error'] = "Слишком много редиректов (> {$this->maxRedirects})";
                        break;
                    }
                } else {
                    $result['status_code'] = $status;
                    $result['final_url'] = $currentUrl;
                    $result['size'] = strlen($response->getBody());
                    break;
                }
            }
            if ($redirectCount > $this->maxRedirects && !$result['error']) {
                $result['error'] = "Превышен лимит редиректов ({$this->maxRedirects})";
            }
        } catch (Exception $e) {
            $result['error'] = $e->getMessage();
        }

        $result['time'] = microtime(true) - $start;
        return $result;
    }

    public function checkMultiple($urls, $workers = 10)
    {
        $results = [];
        // Простая параллельная обработка через массивы (в PHP без многопоточности сложно)
        // Используем последовательную обработку для простоты
        foreach ($urls as $url) {
            $results[$url] = $this->checkUrl($url);
        }
        return $results;
    }
}

function printResults($results)
{
    foreach ($results as $url => $res) {
        echo "\n\033[36m🔍 Проверка: $url\033[0m\n";
        if ($res['error']) {
            echo "\033[31m❌ Ошибка: " . $res['error'] . "\033[0m\n";
            continue;
        }
        $status = $res['status_code'];
        $statusColor = ($status >= 200 && $status < 300) ? "\033[32m" :
                       (($status >= 300 && $status < 400) ? "\033[33m" : "\033[31m");
        $statusText = ($status >= 200 && $status < 300) ? "$status OK" :
                      (($status >= 300 && $status < 400) ? "$status Redirect" : "$status Error");
        echo $statusColor . "✅ " . $res['final_url'] . "\033[0m\n";
        echo "   Статус: " . $statusColor . $statusText . "\033[0m\n";
        printf("   Время: %.2f сек\n", $res['time']);
        echo "   Размер: " . $res['size'] . " байт\n";
        if (count($res['redirects']) > 0) {
            echo "   Редиректы: " . count($res['redirects']) . "\n";
            $i = 1;
            foreach ($res['redirects'] as $r) {
                echo "     $i. " . $r['url'] . " -> " . $r['status_code'] . "\n";
                $i++;
            }
        }
    }
}

function saveJSON($results, $filename)
{
    file_put_contents($filename, json_encode($results, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
    echo "\033[32m💾 Сохранено JSON: $filename\033[0m\n";
}

function saveCSV($results, $filename)
{
    $fp = fopen($filename, 'w');
    fputcsv($fp, ['URL', 'Final URL', 'Status', 'Time (s)', 'Size (bytes)', 'Redirects', 'Error']);
    foreach ($results as $url => $res) {
        fputcsv($fp, [
            $url,
            $res['final_url'] ?? '',
            $res['status_code'] ?? '',
            number_format($res['time'], 3),
            $res['size'] ?? 0,
            count($res['redirects']),
            $res['error'] ?? ''
        ]);
    }
    fclose($fp);
    echo "\033[32m💾 Сохранено CSV: $filename\033[0m\n";
}

function main($argv)
{
    $options = [
        'timeout' => 10,
        'max_redirects' => 10,
        'workers' => 10,
        'headers' => [],
    ];
    $urls = [];

    for ($i = 1; $i < count($argv); $i++) {
        if ($argv[$i] == '--file' || $argv[$i] == '-f') {
            $file = $argv[++$i];
            $lines = file($file, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
            foreach ($lines as $line) {
                $urls[] = trim($line);
            }
        } elseif ($argv[$i] == '--timeout' || $argv[$i] == '-t') {
            $options['timeout'] = (int)$argv[++$i];
        } elseif ($argv[$i] == '--max-redirects' || $argv[$i] == '-m') {
            $options['max_redirects'] = (int)$argv[++$i];
        } elseif ($argv[$i] == '--output-json' || $argv[$i] == '-o') {
            $options['output_json'] = $argv[++$i];
        } elseif ($argv[$i] == '--output-csv' || $argv[$i] == '-c') {
            $options['output_csv'] = $argv[++$i];
        } elseif ($argv[$i] == '--workers' || $argv[$i] == '-w') {
            $options['workers'] = (int)$argv[++$i];
        } elseif ($argv[$i] == '--header' || $argv[$i] == '-H') {
            $h = $argv[++$i];
            $parts = explode(':', $h, 2);
            if (count($parts) == 2) {
                $options['headers'][trim($parts[0])] = trim($parts[1]);
            }
        } elseif (!str_starts_with($argv[$i], '-')) {
            $urls[] = $argv[$i];
        }
    }

    if (empty($urls)) {
        echo "❌ Нет URL для проверки.\n";
        exit(1);
    }

    echo "\033[36m🔗 Link Checker (PHP)\033[0m\n";

    $checker = new LinkChecker($options['timeout'], $options['max_redirects'], $options['headers']);
    $results = $checker->checkMultiple($urls, $options['workers']);
    printResults($results);

    if (isset($options['output_json'])) saveJSON($results, $options['output_json']);
    if (isset($options['output_csv'])) saveCSV($results, $options['output_csv']);
}

$argc = $_SERVER['argc'] ?? 0;
$argv = $_SERVER['argv'] ?? [];
main($argv);
?>
